src/data/buffer/parser/basic.lean

```diff
@@ -2139,4 +2140,221 @@ end
+                 nat.sub_zero],
+      rw [nat.sub_le_sub_right_iff, nat.pred_le_iff],
+      { convert many1_bounded_of_done hp,
+        cases hc : cb.size,
+        { have := bounded.of_done hp,
+          rw hc at this,
+          exact absurd n.zero_le this.not_le },
+        { simp } },
+      { exact nat.le_pred_of_lt (bounded.of_done hp) } },
+  simp [IH _ hdm, hx, hk, rearr, ←chdh, ←ltll, hpow, add_assoc, nat.of_digits_append, mul_comm]
+end
+
+lemma nat_of_done_as_digit {val : ℕ} (h : nat cb n = done n' val) :
+  ∀ (hn : n' ≤ cb.size) k (hk : k < n'), n ≤ k →
+    '0' ≤ cb.read ⟨k, hk.trans_le hn⟩ ∧ cb.read ⟨k, hk.trans_le hn⟩ ≤ '9' :=
+begin
+  simp only [nat, pure_eq_done, and.left_comm, decorate_error_eq_done, bind_eq_done,
+             exists_eq_left, exists_and_distrib_left] at h,
+  obtain ⟨xs, h, -⟩ := h,
+  induction xs with hd tl hl generalizing n n',
+  { simpa using h },
+  cases tl,
+  { simp only [many1_eq_done, many_eq_done_nil, and.left_comm, exists_and_distrib_right,
+               exists_eq_left] at h,
+    obtain ⟨h, -⟩ := h,
+    simp only [digit_eq_done, and.comm, and.left_comm, digit_eq_fail, true_and, exists_eq_left,
+               eq_self_iff_true, exists_and_distrib_left] at h,
+    obtain ⟨rfl, -, hn, ge0, le9, rfl⟩ := h,
+    intros hn' k hk hk',
+    have : k = n := le_antisymm (nat.le_of_lt_succ hk) hk',
+    subst this,
+    exact ⟨ge0, le9⟩ },
+  { rw many1_eq_done at h,
+    obtain ⟨_, hp, h⟩ := h,
+    have := step.of_done hp,
+    subst this,
+    rw ←many1_eq_done_iff_many_eq_done at h,
+    intros hn k hk hk',
+    rcases hk'.eq_or_lt with rfl|hk',
+    { simp [digit_eq_done] at hp,
+      obtain ⟨-, hx, rfl, ge0, le9⟩ := hp,
+      exact ⟨ge0, le9⟩ },
+    { apply hl h,
+      exact nat.succ_le_of_lt hk' } }
+end
+
+lemma nat_of_done_bounded {val : ℕ} (h : nat cb n = done n' val) :
+  ∀ (hn : n' < cb.size), '0' ≤ cb.read ⟨n', hn⟩ → '9' < cb.read ⟨n', hn⟩ :=
+begin
+  simp only [nat, pure_eq_done, and.left_comm, decorate_error_eq_done, bind_eq_done,
+             exists_eq_left, exists_and_distrib_left] at h,
+  obtain ⟨xs, h, -⟩ := h,
+  induction xs with hd tl hl generalizing n,
+  { simpa using h },
+  obtain ⟨k, hk⟩ : ∃ k, cb.size = n + k + 1 := nat.exists_eq_add_of_lt (bounded.of_done h),
+  cases tl,
+  { simp only [many1_eq_done, many_eq_done_nil, and.left_comm, exists_and_distrib_right,
+               exists_eq_left] at h,
+    obtain ⟨-, _, h⟩ := h,
+    simp only [digit_eq_done, and.comm, and.left_comm, digit_eq_fail, true_and, exists_eq_left,
+               eq_self_iff_true, exists_and_distrib_left] at h,
+    obtain (⟨rfl, h⟩ | ⟨h, -⟩) := h,
+    { intro hn,
+      simpa using h hn },
+    { simpa using mono.of_fail h } },
+  { rw many1_eq_done at h,
+    obtain ⟨_, -, h⟩ := h,
+    rw ←many1_eq_done_iff_many_eq_done at h,
+    exact hl h }
+end
+
+lemma nat_eq_done {val : ℕ} : nat cb n = done n' val ↔ ∃ (hn : n < n'),
```

Comment:
As before, do you think you could add some inline comments to make it easier to follow the approach here?